File: src/com/leetcode/medium/graph/RedundantConnection.java
package com.leetcode.medium.graph;

import com.leetcode.utils.ArrayUtils;

/**
 * Created using IntelliJ IDEA.
 * Author:  abhijeet,
 * Date:    06/06/24,
 * Time:    10:44 am
 * 684. Redundant Connection
 * https://leetcode.com/problems/redundant-connection/description/
 */
public class RedundantConnection {
  public static void main(String[] args) {
    RedundantConnection redundantConnection = new RedundantConnection();
    int[][] edges = {{1, 2}, {1, 3}, {2, 3}};
    ArrayUtils.printArray(redundantConnection.findRedundantConnection(edges));
  }

  public int find(int i, int[] parent) {
    if (parent[i] == i) {
      return i;
    }
    return parent[i] = find(parent[i], parent);
  }

  public void union(int i, int j, int[] parent, int[] rank) {
    int parentOfI = find(i, parent);
    int parentOfJ = find(j, parent);
    if (parentOfI == parentOfJ) {
      return;
    }
    if (rank[parentOfI] > rank[parentOfJ]) {
      parent[parentOfJ] = parentOfI;
    } else if (rank[parentOfI] < rank[parentOfJ]) {
      parent[parentOfI] = parentOfJ;
    } else {
      parent[parentOfI] = parentOfJ;
      rank[parentOfI]++;
    }
  }

  private int[] findRedundantConnection(int[][] edges) {
    int size = edges.length;
    int[] parent = new int[size + 1];
    int[] rank = new int[size + 1];
    for (int i = 1; i <= size; i++) {
      parent[i] = i;
      rank[i] = 0;
    }
    for (int[] edge : edges) {
      if (find(edge[0], parent) == find(edge[1], parent)) {
        return edge;
      }
      union(edge[0], edge[1], parent, rank);
    }
    return new int[] {-1, -1};
  }
}

File: leetsync/684-redundant-connection/redundant-connection.go
func find(parent []int, i int) int {
	if parent[i] == i {
		return i
	}
	parent[i] = find(parent, parent[i])
	return parent[i]
}
func union(parent []int, rank []int, i, j int) {
	parentOfI, parentOfJ := find(parent, i), find(parent, j)
	if rank[parentOfI] > rank[parentOfJ] {
		parent[parentOfJ] = parent[parentOfI]
	} else if rank[parentOfI] < rank[parentOfJ] {
		parent[parentOfI] = parent[parentOfJ]
	} else {
		parent[parentOfI] = parent[parentOfJ]
		rank[parentOfJ]++
	}
}

func findRedundantConnection(edges [][]int) []int {
	n := len(edges)
	parent := make([]int, n+1)
	rank := make([]int, n+1)
	for i := 0; i < n; i++ {
		parent[i] = i
		rank[i] = 0
	}
	for _, edge := range edges {
		u, v := edge[0], edge[1]
		if find(parent, u) == find(parent, v) {
			return edge
		}
		union(parent, rank, u, v)
	}
	return []int{}
}